internal/internal_task_handlers.go

```diff
@@ -1028,10 +1028,15 @@ ProcessEvents:
 		}
 		if isReplay {
 			w.workflowInfo.currentTaskBuildID = currentBuildID
-		} else if !isReplay && !isQueryOnlyTask {
-			// Query only tasks should use the build ID from the workflow task, not the worker's
-			// build id, since the user cares about what affected workflow state.
+		} else if !isQueryOnlyTask {
+			// Real new (non-query, non-replay) tasks use the worker's build ID
 			w.workflowInfo.currentTaskBuildID = w.wth.workerBuildID
+		} else {
+			// It *is* a query only task, use the task's build ID (queries are considered non-replay
+			// in go sdk), but only if one was actually discovered.
+			if currentBuildID != "" {
```

Comment:
When is `currentBuildID` empty?

Reply:
It can be empty here if the lookahead didn't find anything, but we wouldn't want to overwrite some previous task which did find something.

Reply:
Yeah I am wondering what scenario look ahead can't find anything?

Reply:
The problem is sort of specific to queries on the first task I think. No WFT completed event exists yet to be looked-ahead-to (well, it does actually in the query, but we don't read it since we're already up-to-date), but the `nextTask` code still runs, which then returns this as empty.

Reply:
Ah I didn't think of the first WFT thanks for clarifying